packages/hierarchies-react/src/presentation-hierarchies-react/internal/UseUnifiedSelection.ts

```diff
@@ -18,97 +17,105 @@ export interface TreeSelectionOptions {
 
 /** @internal */
 export interface UseUnifiedTreeSelectionProps {
+  imodelKey: string;
+  sourceName: string;
   getNode: (nodeId: string) => TreeModelHierarchyNode | TreeModelRootNode | undefined;
 }
 
 /** @internal */
-export function useUnifiedTreeSelection({ getNode }: UseUnifiedTreeSelectionProps): TreeSelectionOptions {
-  const { has, add, remove } = useUnifiedSelectionContext();
+export function useUnifiedTreeSelection({ imodelKey, sourceName, getNode }: UseUnifiedTreeSelectionProps): TreeSelectionOptions {
+  const [options, setOptions] = useState<TreeSelectionOptions>({
+    isNodeSelected: () => false,
+    selectNode: () => {},
+  });
+
+  const selectionStorage = useUnifiedSelectionContext();
+  useEffect(() => {
+    if (!selectionStorage) {
+      setOptions({
+        isNodeSelected: () => false,
+        selectNode: () => {},
+      });
+      return;
+    }
+
+    setOptions(createOptions(imodelKey, sourceName, selectionStorage, getNode));
+    return selectionStorage.selectionChangeEvent.addListener((args) => {
+      if (imodelKey !== args.iModelKey || args.level > 0) {
+        return;
+      }
+
+      setOptions(createOptions(imodelKey, sourceName, selectionStorage, getNode));
+    });
+  }, [selectionStorage, getNode, imodelKey, sourceName]);
 
-  const isNodeSelected = useCallback(
-    (nodeId: string) => {
+  return options;
+}
+
+function createOptions(
+  key: string,
+  source: string,
+  storage: SelectionStorage | undefined,
+  getNode: (nodeId: string) => TreeModelHierarchyNode | TreeModelRootNode | undefined,
+): TreeSelectionOptions {
+  if (!storage) {
+    return {
+      isNodeSelected: () => false,
+      selectNode: () => {},
+    };
+  }
+
+  return {
+    isNodeSelected: (nodeId: string) => {
       const node = getNode(nodeId);
       if (!node || !isTreeModelHierarchyNode(node)) {
         return false;
       }
 
-      return has(getNodeKeys(node.nodeData));
+      const selectables = storage.getSelection(key, 0);
+
+      const hierarchyNode = node.nodeData;
+      if (HierarchyNode.isClassGroupingNode(hierarchyNode)) {
+        return Selectables.has(selectables, { identifier: node.id });
+      }
+
+      if (!HierarchyNode.isInstancesNode(hierarchyNode)) {
+        return false;
+      }
+
+      return Selectables.hasAny(selectables, hierarchyNode.key.instanceKeys);
     },
-    [has, getNode],
-  );
 
-  const selectNode = useCallback(
-    (nodeId: string, isSelected: boolean) => {
+    selectNode: (nodeId: string, isSelected: boolean) => {
       const node = getNode(nodeId);
       if (!node || !isTreeModelHierarchyNode(node)) {
         return;
       }
 
-      const action = isSelected ? add : remove;
-      action(getNodeKeys(node.nodeData));
+      const action = isSelected ? storage.addToSelection.bind(storage) : storage.removeFromSelection.bind(storage);
+      action(source, key, createSelectables(node.id, node.nodeData), 0);
     },
```

Comment:
need to ensure our current uses cases are supported, e.g. the extended selection mode: 
- click to replace, 
- shift+click to invert/select multiple nodes, 
- ctrl+click to (de)select a node.

Reply:
Filled separate issue to avoid expanding scope of this PR. https://github.com/iTwin/presentation/issues/481